scoring/score.py

```diff
@@ -1,14 +1,149 @@
+import collections
+
+NUM_TOKENS_OF_EACH_COLOUR = 10
+ROBOT_MULTIPLIER = 3
+
+
 class InvalidScoresheetException(Exception):
     pass
 
 
 class Scorer(object):
+    # Note: if the arena colours (from arena.yaml) change,
+    # both this and the printed score-sheets will need to change
+    ZONE_COLOURS = [
+        'G',    # zone 0 = green
+        'O',    # zone 1 = orange
+        'P',    # zone 2 = purple
+        'Y',    # zone 3 = yellow
+    ]
+
+    ZONE_SCORE_MAP = [
+        [ 2,  2,  2,  2,  2],
+        [ 2,  7,  7,  7,  2],
+        [ 2,  7, 30,  7,  2],
+        [ 2,  7,  7,  7,  2],
+        [ 2,  2,  2,  2,  2],
+    ]
+
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
-        self._arena_data = arena_data
+        self._zone_contents = arena_data['other']['zone_contents']
+
+        for row in self._zone_contents:
+            for zone in row:
+                zone['tokens'] = collections.Counter(
+                    zone['tokens'].replace(' ', ''),
+                )
+
+        self._colour_to_team = {
+            self.ZONE_COLOURS[info['zone']]: tla
+            for tla, info in self._teams_data.items()
+        }
+
+    def validate(self, extra):
+        tokens = collections.Counter()
+        robots = collections.Counter()
+        for row in self._zone_contents:
+            for zone in row:
+                robots.update(zone['robots'])
+                tokens.update(zone['tokens'])
+
+        extra_robots = set(robots.keys()) - set(self._teams_data.keys())
+        if extra_robots:
+            raise InvalidScoresheetException(
+                "Unepxected robot(s): {!r}".format(extra_robots),
+            )
+
+        for robot, count in robots.items():
+            if count > 1:
+                raise InvalidScoresheetException(
+                    "{} appears {} times! (can only appear once)".format(
```

Comment:
Is this error clear enough that this is an "at most" condition rather than "exactly" condition?

Reply:
08c9f95